Debugger MemoryWidget: Add dual views

TryTwo opened the pull request:
Relies on  #10563.  Restores the dual views ability that WX had, and improves on it.

![PR_dual_view](https://user-images.githubusercontent.com/10532806/162720085-113720ab-8344-4e82-be5c-df48cdfe9d14.jpg)

Looking for input on how customizable it should be. Currently it's always 4 bytes per row and the left side is always hex. It auto chooses the size/columns based on what is on the right.  The second display type box, for bytes per row, doesn't do anything in dual view, because it's always 4 bytes or 8 bytes for a double.

I think the left side always being hex is fine.

Currently Double has a small breakpoint display bug.  I also just added the Hex8/Hex16 ability to the left side, so it got a little more complicated than just having Hex32. I'll maybe need to clean it up more.

Dentomologist commented:
@dolphin-emu-bot rebuild

TryTwo commented:
If I added a hex64 view type the mismatch for Double vs two hex32 columns could be corrected by displaying one hex64 column instead. Hex64 wouldn't need a UI option.  If it's easier to read two hex32's then I'll just leave it with a few special rules to deal with the mismatch.

Dentomologist commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss commented:
You need to rebase this now.

Also, I'm not really happy with this approach. I think it would be way more useful if the bytes-per-row still worked as set, and possibly even to have separate dropdowns for the left and right half's types. Some extra spacing between the two halves couldn't hurt either.

TryTwo commented:
First was rebase. Second is commit for larger size dual views. I'm not sure if you'll like it, so I left it as a second commit. It gets buggy if I have to stretch the table too far, I forget how to fix that atm.

Extra spacing might require an empty column between them.

dreamsyntax commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -179,33 +187,71 @@ void MemoryViewWidget::Update()
+  const AddressSpace::Accessors* accessors = AddressSpace::GetAccessors(m_address_space);
+
+  auto text_alignment = Qt::AlignLeft;
+  if (type == Type::Signed32 || type == Type::Unsigned32 || type == Type::Signed16 ||
+      type == Type::Unsigned16 || type == Type::Signed8 || type == Type::Unsigned8)
+    text_alignment = Qt::AlignRight;
```
Multiline if need braces.

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -91,31 +91,32 @@ static int GetTypeSize(MemoryViewWidget::Type type)
-  case MemoryViewWidget::Type::Unsigned8:
+  case MemoryViewWidget::Type::Hex8:
     return 3;
+  case MemoryViewWidget::Type::Unsigned8:
+    return 4;
   case MemoryViewWidget::Type::Hex16:
   case MemoryViewWidget::Type::Signed8:
-    return 4;
-  case MemoryViewWidget::Type::Unsigned16:
     return 5;
-  case MemoryViewWidget::Type::Signed16:
+  case MemoryViewWidget::Type::Unsigned16:
     return 6;
+  case MemoryViewWidget::Type::Signed16:
+    return 7;
   case MemoryViewWidget::Type::Hex32:
-    return 8;
-  case MemoryViewWidget::Type::Float32:
     return 9;
+  case MemoryViewWidget::Type::Float32:
+    return 11;
   case MemoryViewWidget::Type::Double:
   case MemoryViewWidget::Type::Unsigned32:
   case MemoryViewWidget::Type::Signed32:
-    return 10;
+    return 12;
   default:
-    return 8;
+    return 10;
   }
```
What's up with all the changes here?

sepalani replied:
Can't `std::numeric_limits<T>::digits10` be used here?

TryTwo replied:
Interesting, I'll try that.

I was adding the extra space character into it, so I didn't have to calculate it multiple times. I'll fully rewrite it if numeric limits works.

TryTwo replied:
Documentation says numeric limits will be off-by-one for things like ints, because not all three digit numbers can be represented by an int8, for example.

I could have it give the max characters then go to a calculation to add spacing/padding, if that is clearer. There might be some other way to set the space between columns, but I don't know it.

sepalani replied:
That's how `digits10` [works](https://stackoverflow.com/a/36631469) regarding it being off by one. Here is an example using SFINAE:
```c++
template<typename T,
         typename std::enable_if_t<std::is_integral_v<T>>* = nullptr>
constexpr int character_count()
{
  return 1 + std::numeric_limits<T>::digits10 + int(std::numeric_limits<T>::is_signed);
}

template<typename T,
         typename std::enable_if_t<std::is_floating_point_v<T>>* = nullptr>
constexpr int character_count();

template<>
constexpr int character_count<float>()
{
  // FLT_MAX = 3.40282e+38
  return 11;
}

template<>
constexpr int character_count<double>()
{
  // DBL_MAX = 1.79769e+308
  return 12;
}
```
I don't think we gain much using this approach, though.

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -295,29 +341,53 @@ void MemoryViewWidget::Update()
-  int max_length = GetCharacterCount(m_type);
+  for (int i = 0; i < rowCount(); i++)
+  {
+    bool row_breakpoint = true;
```
Not really your fault, but this is a bit broken, eg. set a row breakpoint on 16 bytes per row, then switch to 4 bytes per row and try to unset them. I'd just remove the row breakpoint altogether and just keep the red backdrop and right-click menu breakpoints, but ymmv.

TryTwo replied:
Hmm I could set/unset multiple breakpoints according to row length + type size.

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -295,29 +341,53 @@ void MemoryViewWidget::Update()
+      // In dual view the only sizes that dont match up on both left and right views are for
+      // Double, which uses two or four columns of hex32.
+      int type_size = GetTypeSize(m_type);
+      if (m_dual_view && m_type == Type::Double && c < (columnCount() == 8 ? 6 : 4))
```
This condition on the right there is magic number nonsense. There's gotta be a better way to express this.

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryWidget.cpp:
```diff
@@ -431,6 +436,10 @@ void MemoryWidget::OnDisplayChanged()
   int alignment;
+  bool dual_view = m_dual_check->isChecked();
+
+  // if (dual_view)
+  //   bytes_per_row = 4;
```
Commented out code.

AdmiralCurtiss reviewed (commented):
A few nitpicks but otherwise this is fine. I'd still like to see a bit of a spacing between the two halves, and the ability to have the same hex view on both sides is kinda weird, though with the UI as-is it makes sense I suppose.

TryTwo commented:
Trying to just use Hex64 when dual view + double is selected. 
Removed Hex + Hex dual view.
Added some spacing for dual view.

Should I have it save all the display type settings between sessions?

sepalani commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -91,31 +92,34 @@ static int GetTypeSize(MemoryViewWidget::Type type)
 
 static int GetCharacterCount(MemoryViewWidget::Type type)
```
If you plan to include the space, what about `GetCellSize` as a function name? Can't this function be marked as `constexpr`. I also highly suggest that for each case, you write an example.

For instance:
```c++
case MemoryViewWidget::Type::Unsigned32: // UINT_MAX = 4294967295
  return 11;
case MemoryViewWidget::Type::Float32: // FLT_MAX = 3.40282e+38
case MemoryViewWidget::Type::Signed32: // INT_MIN = -2147483648
  return 12;
```

TryTwo replied:
Yeah I don't think it will be used for anything else. I can add examples too.

TryTwo replied:
Actually the cell size needs to be calculated from the font, so I think CharacterCount is most accurate.

TryTwo commented:
Tried a solution to row breakpoints.

sepalani commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss commented:
I can still get the row breakpoints to break with this. But again, I wouldn't worry about it, it can be fixed or removed in a separate PR. Your commits are also now a combined thing of multiple features in one commit, which is far worse than row breakpoints being screwy.

AdmiralCurtiss commented:
I would also argue that the Dual View checkbox not doing anything in Hex view is worse than it displaying two Hex views, from a discoveryability perspective.

TryTwo commented:
My goal isn't to change the features of row breakpoints, but to keep them from breaking.  I think some of the breaking happened in the previous PR though, so I can move the changes to a new PR if you want.  Either way I think I got them fixed.

sepalani commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss reviewed (approved):
Yeah I'm okay with this now.
